draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,937 @@
+    status.
+
+  In addition, the actual request performed in the task must also stand in the
+  aforementioned conditions. Otherwise, the health check will change it's state.
+
+  The health check can be in an **Insecure** state if one or more of the
+  following conditions are met:
+
+  * The TLS certificates are invalid for any reason.
+  * The Intrusion Detection System has reported that the network is compromised
+    for any reason.
+
+  It is up for the user to configure the :term:`Circuit Breaker` to prevent
+  insecure requests from being executed.
+
+  The health check can be in an **Degraded** state if one or more of the
+  following conditions are met:
+
+  * The request does not reply with a 2xx HTTP status.
+  * The request responds slowly and almost reaches it's time limits.
+
+  It is up for the user to configure the :term:`Circuit Breaker` to prevent
+  requests from being executed after multiple attempts or not all.
+
+  The health check can be in an **Unhealthy** state if one or more of the
+  following conditions are met:
+
+  * The request responds with a 500 HTTP status.
+  * The request's response has not been received within the specified time
+    limits.
+
+  It is up for the user to configure the :term:`Circuit Breaker` to prevent
+  requests from being executed if there is an issue with the endpoint.
+
+  The health check can be in an **Permanently Unavailable** state if one or more
+  of the following conditions are met:
+
+  * The request responds with a
+    `404 Not Found <https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404>`_
+    HTTP status.
+  * The HTTP method we're about to use is not allowed.
+  * The content type we're about to use is not allowed.
+
+Circuit Breaking
+~~~~~~~~~~~~~~~~
+
+Celery 5 introduces the concept of :term:`Circuit Breaker` into the framework.
+
+A Circuit Breaker prevents a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+or a :ref:`service <draft/celery-5-high-level-architecture:Services>`
+from executing.
+
+Each task or a service has a Circuit Breaker which the user can associate
+health checks with.
+
+In addition, if the task or the service unexpectedly fails, the user
+can configure the Circuit Breaker to trip after a configured number of times.
+The default value is 3 times.
+
+Whenever a Circuit Breaker trips, the worker will emit a warning log message.
+
+After a configured period of time the circuit is opened again and tasks may
+execute. The default period of time is 30 seconds with no linear or exponential
+growth.
+
+The user will configure the following properties of the Circuit Breaker:
+
+* How many times the health checks may fail before
+  the circuit breaker trips.
+* How many unexpected failures the task or service tolerates before tripping
+  the Circuit Breaker.
+* The period of time after which the circuit is yet
+  again closed. That time period may grow linearly or exponentially.
```

Comment:
(In relation to health checks) Most implementations I've seen of this concept will only become "active" again after a set number of *healthy* checks, not a particular time.